[c10d] Enable gather_object on nccl

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71623
* #70029
* #66745

Enable gather_object on the nccl backend, since we already support `dist.gather` on nccl. This requires user to set the current device properly.

Differential Revision: [D33701042](https://our.internmc.facebook.com/intern/diff/D33701042/)

Commit:
[c10d] Enable gather_object on nccl

Enable gather_object on the nccl backend, since we already support `dist.gather` on nccl. This requires user to set the current device properly.

Differential Revision: [D33701042](https://our.internmc.facebook.com/intern/diff/D33701042/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5530,6 +5530,17 @@ def test_all_gather_object_subgroup(self):
+
+            # If GPU test, add object with GPU tensor
+            if backend == "nccl":
+                gather_objects.append(Foo(torch.randn(3, 3, device=0)))
```
should the gpu tensor be allocated on the current device instead of 0 always?

rohan-varma reviewed (approved):
LGTM, just one question

Commit:
Update on "[c10d] Enable gather_object on nccl"


Enable gather_object on the nccl backend, since we already support `dist.gather` on nccl. This requires user to set the current device properly.

Differential Revision: [D33701042](https://our.internmc.facebook.com/intern/diff/D33701042/)

[ghstack-poisoned]

Commit:
Update on "[c10d] Enable gather_object on nccl"


Enable gather_object on the nccl backend, since we already support `dist.gather` on nccl. This requires user to set the current device properly.

Differential Revision: [D33701042](https://our.internmc.facebook.com/intern/diff/D33701042/)

[ghstack-poisoned]

Commit:
Update on "[c10d] Enable gather_object on nccl"


Enable gather_object on the nccl backend, since we already support `dist.gather` on nccl. This requires user to set the current device properly.

Differential Revision: [D33701042](https://our.internmc.facebook.com/intern/diff/D33701042/)

[ghstack-poisoned]
